### ISSUE TITLE
Database checks to prevent high cost, transaction data corruption

### ISSUE THREAD

**Issue by petmongrels** [avniproject/avni-server#645]
There are certain data errors whose cost is very high and getting to know them via alerts doesn't reduce the cost that much. These errors likely related to transaction data. We can make use of custom database constraints to prevent them then, handle them after they have happened. Some custom database constraints cannot be applied at query level, for these we can spike and use DEFERRED constraints feature in Postgres (deferred to commit time).

These are the checks that we should have in our system as their failure can lead high cost.

- Sync attribute consistency across TX entities
- Address id consistency across TX entities
- Subject migration timestamps older than entity timestamp, across TX entities

**Comment by mahalakshme** [comment]
@petmongrels 
Postgres documentation says:  https://www.postgresql.org/docs/current/sql-set-constraints.html
```
Currently, only UNIQUE, PRIMARY KEY, REFERENCES (foreign key), and EXCLUDE constraints are affected by this setting. NOT NULL and CHECK constraints are always checked immediately when a row is inserted or modified (not at the end of the statement). Uniqueness and exclusion constraints that have not been declared DEFERRABLE are also checked immediately.
```
So looks like db constraint might not be possible, but may be we can use Spring AOP for this validation - but since again this will be part of validation at code level not sure if it will be that effective.

**Comment by petmongrels** [comment]
Yes function based constraint won't work for this kind of use cases where multiple entities get updated together.

But we can do it using Hibernate postFlush, beforeTransactionCompletion methods of hibernate interceptor and thread-local. Persistence via hibernate is the most difficult one to ensure database integrity for. The way we enforce constraint should be via direct SQL - so that it is fast. Hibernate interceptors is only to know what has changed and get a callback before commit.

I have also added individual_id enforcing.
